InnerEye/ML/SSL/utils.py

```diff
@@ -0,0 +1,99 @@
+#  Copyright (c) Microsoft Corporation. All rights reserved.
+#  Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+#  ------------------------------------------------------------------------------------------
+
+import logging
+from enum import Enum
+from pathlib import Path
+from typing import Optional
+
+import torch
+
+from InnerEye.ML.SSL import ssl_augmentation_config
+from InnerEye.ML.SSL.config_node import ConfigNode
+from InnerEye.ML.SSL.encoders import DenseNet121Encoder
+from InnerEye.ML.lightning_container import LightningModuleWithOptimizer
+
+
+class SSLModule(Enum):
+    ENCODER = 'encoder'
+    LINEAR_HEAD = 'linear_head'
+
+
+class SSLType(Enum):
+    SimCLR = "SimCLR"
+    BYOL = "BYOL"
+
+
+def load_ssl_model_config(config_path: Path) -> ConfigNode:
+    """
+    Loads configs required for self supervised learning. Does not setup cudann as this is being
+    taken care of by lightning.
+    """
+    config = ssl_augmentation_config.get_default_model_config()
+    config.merge_from_file(config_path)
+    config.freeze()
+    return config
+
+
+def create_ssl_encoder(encoder_name: str, dataset_name: Optional[str] = None) -> torch.nn.Module:
+    """
+    Creates SSL encoder.
+    :param encoder_name: available choices: resnet18, resnet50, resnet101 and densenet121.
+    :param dataset_name: optional, if "CIFAR10" the initial convolution layer with be adapted to not shrink the
+    images. Else if None or other the argument is ignored.
+    """
+    from pl_bolts.models.self_supervised.resnets import resnet18, resnet50, resnet101
+    if encoder_name == 'resnet18':
+        encoder = resnet18(return_all_feature_maps=False)
+    elif encoder_name == 'resnet50':
+        encoder = resnet50(return_all_feature_maps=False)
+    elif encoder_name == 'resnet101':
+        encoder = resnet101(return_all_feature_maps=False)
+    elif encoder_name == 'densenet121':
+        encoder = DenseNet121Encoder()
+    else:
+        raise ValueError("Unknown model type")
+
+    if dataset_name is not None and dataset_name.startswith("CIFAR"):
+        logging.info("Updating the initial convolution in order not to shrink CIFAR10 images")
+        encoder.conv1 = torch.nn.Conv2d(3, 64, kernel_size=3, stride=1, padding=1, bias=False)
+
+    return encoder
+
+
+def create_ssl_image_classifier(num_classes: int, freeze_encoder: bool, pl_checkpoint_path: str,
+                                class_weights: Optional[torch.Tensor] = None) -> LightningModuleWithOptimizer:
+    """
+    Creates a SSL image classifier from a frozen encoder trained on in an unsupervised manner.
+    """
+    from InnerEye.ML.SSL.byol.byol_module import BYOLInnerEye
+    from InnerEye.ML.SSL.simclr_module import SimCLRInnerEye
+    from InnerEye.ML.SSL.ssl_online_evaluator import WrapSSL
+    from InnerEye.ML.SSL.lightning_containers.ssl_image_classifier import SSLClassifier
```

Comment:
Is there a reason why these imports are local? If so, maybe add a comment?

Reply:
Yes, circular imports SimCLRInnerEye imports `SSLModuleType` from the same file